src/cocotb/_scheduler.py

```diff
@@ -586,17 +584,18 @@ def _queue(
         """Queue *task* for scheduling.
 
         If the task is already scheduled and you are attempting to reschedule it with a different outcome,
-        we assume that is a mistake and raise an Exception.
-        This behavior may change in the future.
+        we unschedule
         """
         # Don't queue the same task more than once (gh-2503)
         if task in self._pending_tasks:
-            if self._pending_tasks[task] != outcome:
-                raise InternalError(
-                    f"Attempted rescheduling {task!r} with different outcome!"
+            if _debug:
+                self.log.debug(
+                    f"Rescheduling {task!r} with {outcome!r}. Previously scheduled with {self._pending_tasks[task]}"
                 )
-            elif _debug:
-                self.log.debug(f"Not rescheduling already scheduled {task!r}")
+            self._unschedule(
+                task
+            )  # this won't work currently =/ _unschedule does a lot more than it really should.
```

Comment:
This will work once `_unschedule` is refactored after `cancel` change comes in. We won't have to worry about scheduler reentrancy after that so we can move the test end checking stuff out of `_unschedule` into `_schedule` or something since that will be the only location that can call `_unschedule`.